lib/ngx/re.lua

```diff
@@ -0,0 +1,74 @@
+-- Copyright (C) Thibault Charbonnier (thibaultcha)
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local re_match_helper = core_regex.re_match_helper
+local tostring = tostring
+local new_tab = base.new_tab
+local sub = string.sub
+
+
+local _M = { version = base.version }
+
+
+function _M.split(subj, regex, opts, ctx, max, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = new_tab(0, 1)
+    end
+
+    if not ctx.pos then
+        ctx.pos = 1
+    end
+
+    if not res then
+        res = new_tab(0, 0)
+    elseif type(res) ~= "table" then
+        return error("res is not a table", 2)
+    end
+
+    local sub_idx = ctx.pos
+    local res_idx = #res + 1
```

Comment:
This is wrong since the user-supplied `res` table might not be cleared. Also, I think we should just initialize `res_idx` to zero. And everytime we want to append a new element, we always *first* increment the `res_idx` variable.